app/provider_details/provider_service.py

```diff
@@ -36,42 +42,79 @@ def init_app(
-        template_or_service_provider_id = self._get_template_or_service_provider_id(notification)
-        if template_or_service_provider_id:
-            provider = get_provider_details_by_id(template_or_service_provider_id)
+        """
+        Return an instance of ProviderDetails that is appropriate for the given notification.
+        """
 
-            if provider is None:
-                raise InvalidProviderException(f'provider {template_or_service_provider_id} could not be found')
-            elif not provider.active:
-                raise InvalidProviderException(f'provider {template_or_service_provider_id} is not active')
+        # This is a UUID (ProviderDetails primary key).
+        provider_id = self._get_template_or_service_provider_id(notification)
+        logger.debug("notification = %s", notification)
+        logger.debug("provider_id = %s", provider_id)
 
-        else:
-            provider_selection_strategy = self._strategies[NotificationType(notification.notification_type)]
-            provider = provider_selection_strategy.get_provider(notification)
+        if provider_id:
+            provider = get_provider_details_by_id(provider_id)
+        elif notification.notification_type != NotificationType.SMS:
+            # Use an alternative strategy to determine the provider.
+            provider_selection_strategy = self._strategies.get(NotificationType(notification.notification_type))
+            logger.debug("Provider selection strategy: %s", provider_selection_strategy)
+            provider = None if (provider_selection_strategy is None) \
+                else provider_selection_strategy.get_provider(notification)
 
-            if provider is None:
+            if provider is None and provider_selection_strategy is not None:
+                # This exception message is more detailed than the messages below.
                 raise InvalidProviderException(
-                    f'provider strategy {provider_selection_strategy.get_label()} could not find a suitable provider'
+                    f"Could not determine a provider using strategy {provider_selection_strategy.get_label()}."
                 )
+        else:
+            # Do not use any other criteria to determine the provider for SMS notifications.
+            # Unlike e-mail providers, which are basically fungible, SMS providers have more specific
+            # limitations that should preclude selecting different ones in an ad-hoc manner.
+            provider = None
+
+        if provider is None:
+            if provider_id:
+                raise InvalidProviderException(f"The provider {provider_id} could not be found.")
+            raise InvalidProviderException("Could not determine a provider.")
+        elif not provider.active:
+            raise InvalidProviderException(f"The provider {provider.display_name} is not active.")
 
+        logger.debug("Returning provider: %s", "None" if (provider is None) else provider.display_name)
```

Comment:
```suggestion
        logger.debug("Returning provider: %s", provider if provider is None else provider.display_name)
```
No need to specify "None", that's just how it works if it's None. Could  specify `None` without the quotes too.

Reply:
Having `None` there makes it more clear imo. I'll remove the quotes though, since it doesn't need to be a string.